Integrate new collections [ci: last-only]

lrytz opened the pull request.

pathikrit committed:
First commit of ArrayDeque for collections-strawman

szeiger committed:
More minor bug and compatibility fixes from my integration branch

szeiger committed:
immutable.Queue from the old collections library

szeiger committed:
mutable.PriorityQueue from the old collections

We don’t have ResizableArray anymore so we use ArrayBuffer instead.
It wasn’t clear to me why PriorityQueue didn’t use ArrayBuffer in the
first place rather than a custom implementation of ResizableQueue.

szeiger committed:
mutable.ListMap from the old collections library

julienrf committed:
- Make “length” a (final) alias for “size”
- Implementors of indexed collections implement a “finiteSize” member.

julienrf committed:
Make ArrayDeque compile

julienrf committed:
Add a JMH benchmark for ArrayDeque

szeiger committed:
Generalize LazyList to share the implementation with Stream

- Smoke tests for LazyList are disabled for now due to
  https://github.com/lampepfl/dotty/issues/3919. We still run the main
  test suite (but not on Dotty)

julienrf committed:
Introduce Cloneable trait

julienrf committed:
Merge pull request #373 from lrytz/buildFromNotFound

implicitNotFound for BuildFrom

julienrf committed:
Merge pull request #369 from julienrf/cloneable

Introduce Cloneable trait

julienrf committed:
Merge pull request #371 from julienrf/string-array-implicit-factories

Add Factory and BuildFrom instances for String and Array

julienrf committed:
Merge pull request #368 from julienrf/clean-warnings

Address warnings

szeiger committed:
The UnrolledBuffer type from the old collections

msteindorfer committed:
Reimplementations of immutable HashSet and HashMap.

The reimplementations are based upon Compressed Hash-Array Mapped Prefix-trees (CHAMP), see paper "Optimizing Hash-Array Mapped Tries for Fast and Lean Immutable JVM Collections" by Steindorfer and Vinju (OOPSLA'15) for more details and descriptions of low-level performance optimizations (a pre-print of the paper is available under https://michael.steindorfer.name/publications/oopsla15.pdf). This commit closes scala/collection-strawman#192.

The new implementations (i.e., ChampHashSet and ChampHashMap) currently exist next to the previous HashMap and HashSet. By default immutable.Map and immutable.Set now pickup the CHAMP data structures. A JVM flag (-Dstrawman.collection.immutable.useBaseline=true) allows to switch back to the previous HashSet and HashMap implementations for testing. Note, the flag and the previous HashSet and HashMap implementations will be removed in the final version of collection-strawman, but for the time being they remain to support comparing the different trade-offs and performance characteristics of the current and the new data structures.

Preliminary performance numbers of the new CHAMP data structures were presented in issue scala/collection-strawman#192. Overall one can summarize that the CHAMP data structures significantly lower memory footprints and significantly improve all iteration-based operations and equality checks. Basic operations such as lookup, insertion, and deletion may slow down. The current state of the reimplementation does not optimize for hash-collisions yet.

Note that the CHAMP design / implementation differs from the previous immutable hashed data structures by not memoizing the hash codes of the individual elements (which may change the performance of certain workloads). If necessary, CHAMP's design allows to modularly add memoized hash codes of the individual elements (at the expense of some memory savings). Details are discussed in the paper mentioned above.

szeiger committed:
Add search methods from scala.collection.Searching

It is not clear from https://github.com/scala/scala/pull/903 why these
methods were not added directly to the collection classes, so that is
what I’m doing here.

`Searching` still exists and contains `SearchResult` and its subtypes.
We should consider moving `SearchResult` directly into the package
and `Found` and `InsertionPoint` into its companion object.

julienrf committed:
Set version to 0.9.0

julienrf committed:
Set version to 0.10.0-SNAPSHOT

szeiger committed:
Merge pull request #384 from scala/wip/searching

Add search methods from scala.collection.Searching

marcelocenerine committed:
Overload unzip in ArrayOps

marcelocenerine committed:
Fix NPE in Iterator.addString when there are null elements

szeiger committed:
OpenHashMap from the old collections, now deprecated

julienrf committed:
Merge pull request #387 from marcelocenerine/npe_iterator_addString

Fix NPE in Iterator.addString when there are null elements

marcelocenerine committed:
Fix serialization issues in HashMap, HashSet, LinkedHashMap and LinkedHashSet

marcelocenerine committed:
Make immutable.TreeSet serializable

szeiger committed:
Merge pull request #388 from scala/wip/openhashmap

OpenHashMap from the old collections, now deprecated

szeiger committed:
Memoize IterableOnce source on first View traversal

This allows us to get rid of `IterableFactoryLike` and still define
`View.from` in a reasonable way.

szeiger committed:
First-class treatment of ClassTag-based factories

This reinstates the factory part of the abstraction for arbitrary
evidence type constructors which we previously abandoned in favor of
SortedIterableFactory. Now we have a good reason for an
EvidenceIterableFactory with several Ordering-based and several
ClassTag-based factory instances.

ImmutableArray and WrappedArray now have a ClassTag-based factory in
their companion object so that building unboxed arrays with a proper
element type is always the default (and used automatically in calls like
`xs.to(ImmutableArray)` which would previously go through the regular
IterableFactory and build a boxed representation). IterableFactories are
still available as `{ImmutableArray/WrappedArray}.untagged`.

There are no new abstractions at the collection level. We could
consider reintroducing a ConstrainedIterable but at lower levels of
the hierarchy it doesn’t make sense because the Ordering- and
ClassTag-based collections are of very different types.

There are also some fixes for bad arraycopy calls (between boxed and
unboxed representations) and unnecessary boxing in combinators in
ImmutableArray.

szeiger committed:
Merge pull request #361 from scala/wip/stream

Generalize LazyList to allow implementation of Stream

marcelocenerine committed:
Fix bug in indexOfSlice/lastIndexOfSlice when Seq.size == slice.size

marcelocenerine committed:
Bring back collection.convert.ImplicitConversions

julienrf committed:
Merge pull request #393 from marcelocenerine/buggy_indexOfSlice

[392] Fix bug in indexOfSlice/lastIndexOfSlice when Seq.size == slice.size

marcelocenerine committed:
Fix StackOverflowError in Java wrappers' iterator()

marcelocenerine committed:
Make LazyList.drop safe

marcelocenerine committed:
Make Buffer implement SeqOps to retain original type

marcelocenerine committed:
Increment end in Buffer.insert

julienrf committed:
Merge pull request #394 from marcelocenerine/implicitConversions

Bring back collection.convert.ImplicitConversions and fix StackOverflowError in Java wrappers' iterator()

julienrf committed:
Merge pull request #417 from marcelocenerine/buffer_insert

Increment end in Buffer.insert

julienrf committed:
Merge pull request #416 from marcelocenerine/buffer_losing_type

Make Buffer implement SeqOps to retain original type after invoking operations on it

julienrf committed:
Merge pull request #386 from marcelocenerine/array_unzip

Overload unzip in ArrayOps

szeiger committed:
Merge pull request #391 from scala/wip/classtagiterablefactory

First-class treatment of ClassTag-based factories

lrytz committed:
Merge pull request #389 from marcelocenerine/not_serializable

Fix serialization issues in HashMap, LinkedHashMap, HashSet and LinkedHashSet

julienrf committed:
Merge pull request #342 from msteindorfer/new-immutable-hash-set-and-map

New immutable hash set and map

szeiger committed:
Merge pull request #383 from scala/wip/unrolledbuffer

The UnrolledBuffer type from the old collections

julienrf committed:
Merge pull request #415 from marcelocenerine/unsafe_lazylist_drop

Make LazyList.drop safe

marcelocenerine committed:
Fix ListBuffer.filterInPlace

marcelocenerine committed:
Make Buffer.takeInPlace safe

marcelocenerine committed:
Fix ArrayBuffer.filterInPlace

marcelocenerine committed:
Make behavior of ListBuffer.remove(i, n) consistent with ArrayBuffer when n < 0

marcelocenerine committed:
Fix ListBuffer.remove(idx)

marcelocenerine committed:
Make Buffer.dropInPlace safe

marcelocenerine committed:
Make Buffer.dropRightInPlace safe

foobatman committed:
#226 Fix withDefault operations on SortedMap (#381)

#226 Fixes withDefault operations on SortedMap

marcelocenerine committed:
Make Buffer.takeRightInPlace safe

marcelocenerine committed:
Make Buffer.trimStart/trimEnd safe

marcelocenerine committed:
Write unit tests for Buffer.takeWhileInPlace/dropWhileInPlace

marcelocenerine committed:
Fix ArrayBuffer/ListBuffer flatMapInPlace

marcelocenerine committed:
Fix Vector.prependedAll (#414)

julienrf committed:
Merge pull request #423 from marcelocenerine/bug_fixes_buffer

Bug fixes for Buffer/ArrayBuffer/ListBuffer

julienrf committed:
Merge pull request #424 from marcelocenerine/fix_buffer_flatMapInPlace

Fix ArrayBuffer/ListBuffer flatMapInPlace

lrytz committed:
Specific C type parameter for StringBuilder

Also sdd methods to StringBuilder that exist in 2.12.

julienrf committed:
Add old mutable.MultiMap
Rename MultiMap (in contrib) to MultiDict

szeiger committed:
Merge pull request #336 from scala/wip/queue

A port of mutable.Queue from the old collections

szeiger committed:
Fix problems caused by merging https://github.com/scala/collection-strawman/pull/336

The branch was outdated without causing proper merge conflicts.

szeiger committed:
Merge pull request #429 from scala/wip/queue-fix

Fix problems caused by merging #336

julienrf committed:
Merge pull request #357 from julienrf/canbuildfrom

Add deprecated CanBuildFrom alias to BuildFrom

julienrf committed:
Merge pull request #356 from julienrf/final-length

Standardize on .size (but still offer .length)

julienrf committed:
Move IndexedView in its own file

julienrf committed:
Fix compilation errors

julienrf committed:
Relax type bounds

Remove some upper bound constraints on Ops traits.
Remove unnecessary ArrayLike trait.

marcelocenerine committed:
Drop validation in View.Patched so that Buffer.patch does not throw IOOBE

marcelocenerine committed:
Fix ArrayBuffer/ListBuffer.patchInPlace

julienrf committed:
Add SeqView. Make sure getting a View’s View is a no-op.

marcelocenerine committed:
ArrayBuffer.apply/update to ensure index is within bounds

julienrf committed:
Add MapView and make View.MapValue a MapView.

julienrf committed:
Optimize knownSize of small Maps

marcelocenerine committed:
Make sure ArrayBuffer.clear dereference removed elements

julienrf committed:
Failed attempt to fix Dotty compilation error

julienrf committed:
Merge pull request #427 from julienrf/deprecated-multimap-bis

Add old mutable.MultiMap

lrytz committed:
Add deprecated immutable.TreeMap.insert

lrytz committed:
Clarifications for className

julienrf committed:
Implement some IndexedViews in terms of SeqViews

julienrf committed:
Merge pull request #434 from marcelocenerine/buffer_fixes

Bunch of bug fixes for `ArrayBuffer` & `ListBuffer`

julienrf committed:
Add back optimizations of HashSet

julienrf committed:
Add hashset benchmarks

szeiger committed:
Merge pull request #430 from scala/wip/priorityqueue

mutable.PriorityQueue from the old collections

marcelocenerine committed:
Set @SerialVersionUID to 3L for all collections

marcelocenerine committed:
Add missing @SerialVersionUID to serializable collections

szeiger committed:
Merge pull request #428 from scala/wip/memoize-iterableonce

Memoize IterableOnce source on first View traversal

szeiger committed:
Revert https://github.com/scala/collection-strawman/pull/356

Commit:
Make toString in LazyList consistent with other collections.

Fixes scala/collection-strawman#376
Fixes scala/collection-strawman#420

lrytz committed:
deprecated mutable.IterableOps.transform

szeiger committed:
Remove ArrayLike

lrytz committed:
Iterator toString

szeiger committed:
Merge pull request #443 from scala/wip/revert-indexed-length

Revert length / size changes

lrytz committed:
override sizeHint in Builder.mapResult

lrytz committed:
include index in IOOBE

lrytz committed:
Iterator.copyToArray

szeiger committed:
Merge pull request #448 from scala/wip/immutable-queue

immutable.Queue from the old collections library

szeiger committed:
Merge pull request #449 from scala/wip/mutable-listmap

mutable.ListMap from the old collections library

julienrf committed:
Merge pull request #439 from julienrf/hashset-subsetof

Add back optimizations of HashSet

julienrf committed:
Merge pull request #444 from marcelocenerine/serial_version_uid

Harmonize use of @SerialVersionUID

lrytz committed:
IsIterableLike for array

Otherwise the pattern matcher crashes when compiling
value-class-extractor-seq.scala

lrytz committed:
ArrayOps zipWithIndex overload

lrytz committed:
Array.transpose

https://github.com/scala/collection-strawman/issues/435

julienrf committed:
Fix snapshots publication

lrytz committed:
Add `array.deep` extension method to partest-extras (used in tests)

lrytz committed:
partest --pos passes

lrytz committed:
partest --neg passes

julienrf committed:
Update the status and roadmap

julienrf committed:
Fix versions in setup instructions

julienrf committed:
Move new collections and operations to the Wiki

julienrf committed:
Remove the unmaintained list of supported collections and operations

szeiger committed:
Merge pull request #447 from scala/wip/compat-methods3

More minor bug and compatibility fixes from my integration branch

SethTisue committed:
fix typo in .travis.yml

perhaps this will make snapshots start publishing themselves again

references #451

julienrf committed:
Merge pull request #453 from SethTisue/fix-travis-config

fix typo in .travis.yml

kelebra committed:
Added type parameter to Iterator.empty and test cases (#450)

* Added type parameter to Iterator.empty and test cases

* Added extension of IterableFactory in Iterator's companion object

* Added missing final keyword

* Changes according to review in tests.

* Changes according to review in tests

* Changes according to review in tests

jiminhsieh committed:
Bump sbt version & cache depdencies (#455)

* Fix typo.

* Bump sbt version.

* Cache for Travis CI.

tim-ruhland committed:
add the bare and separator-only variants of addString to Iterator

addresses #401

szeiger committed:
Remove the boilerplate associated with the new design (and more)

- The respective `CC` types are exposed (to `protected[this]`) through
  aliases `IterableCC`, `SortedIterableCC`, `MapCC` and `SortedMapCC`
  in the respective `Ops` traits, plus a `BitSetC` for the shared
  `BitSetOps` trait.

- This allows defining `fromSpecificIterable`, `newSpecificBuilder`
  and `empty` (for all but `Iterable`) in the respective collection
  traits `Iterable`, `SortedSet`, `Map` and `SortedMap` (*not* in their
  `Ops` traits) where the `CC` type is visible and set to a concrete
  type but can still be refined in subtypes. This gives us a valid
  implementation until a concrete collection type with a refined `CC`
  gets defined. In this case the inherited implementations have the
  wrong type, so the user is forced to override them.

- Implementations of `fromSpecificIterable`, `newSpecificBuilder` and
  `empty` can be removed from almost all collection implementations
  except the ones where they need to be refined even further than what
  a factory can provide (e.g. `IntMap`, `PriorityQueue`).

- Factories are generally overridden in abstract collection types that
  refine a `CC` (e.g. `Iterable` -> `Seq` -> `immutable.Set` ->
  `immutable.IndexedSeq`) so that concrete implementations without
  further refinement do not need to override them.

- DefaultMap is deprecated because there is no more boilerplate left
  that it could implement.

- `sortedFromIterable`, `mapFromIterable` and `sortedMapFromIterable`
  are treated the same way as `fromIterable`. They are implemented as
  `inline final` calls to the respective factory method.

Fixes https://github.com/scala/collection-strawman/issues/335

szeiger committed:
Remove mapFromIterable and sortedMapFromIterable

The scaladoc comment claims that `mapFromIterable` is “similar to
fromIterable” but that method doesn’t exist (anymore). Unlike
`fromSpecificIterable`, `mapFromIterable` is not specific to the
source element type and therefore can never do more than
`mapFactory.from`. The same goes for `sortedMapFromIterable`.

cryptom-dev committed:
Fix slice behavior difference of ImmutableArray

szeiger committed:
Merge pull request #458 from scala/wip/remove-mapfromiterable

Remove mapFromIterable and sortedMapFromIterable

szeiger committed:
Merge pull request #463 from ackratos/wip/arraySliceException

Fix slice behavior difference of ImmutableArray

julienrf committed:
Merge pull request #452 from lrytz/assorted

Small improvements from my integration branch

julienrf committed:
Take advantage of IndexedViews in IndexedSeqOps

Add IndexedView.Slice
Reduce indirection levels by generalizing views to accept XxxOps
parameters instead of collection types
Add MapView.FilterKeys

julienrf committed:
Remove temporary hack

szeiger committed:
Merge pull request #468 from scala/wip/default-iterable-factory

Remove the boilerplate associated with the new design (and more)

szeiger committed:
Fix the iterableFactory of mutable Sets

This points to a soundness bug in the compiler. These errors should be
caught. Omitting a necessary override leads to the expected errors in
some cases but not in all.

szeiger committed:
Merge pull request #470 from scala/wip/fix-set-factories

Fix the iterableFactory of mutable Sets

szeiger committed:
Make deprecated + work for generic Set and Map

We need to make Set.+ invariant to get the right (2.12-compatible)
behavior in SortedSet.

The deprecated collection.Map.+ was missing an override in
collection.SortedMap to get the correct result (like it was already
done for the non-deprecated immutable versions).

szeiger committed:
Use WrappedArray’s iterableFactory for ArraySeq

The old AraySeq was never specialized, so this is the closest
equivalent that we have.

optician committed:
LazyList.force returns LazyList.  (#454)

* * Add tests of `force` for LazyList and Stream.
* Test string representations of LazyList and Stream after `force` call.
* Add tests of `sameElements` for LazyList and Stream.
* Change `force` result type from `Option[(head, tail)]` to `CC[A]`.
* Force `force` to evaluate whole LazyList and Stream.
* Update methods depended on `force`: `unapply` and `sameElements`.

* Reimplement `sameElements`.
** Remove Tuple allocation.
** Check reference equality on every step.

* * Code style.

* * Remove redundant override of `sameElements` method.

* * Set more strict type of `force` method.

* * Stack safe `force` with cyclic references detection.

* * Update LazyListForceBenchmark

* * Fix method name.
* Test `sameElements` with cyclic collections.
* Reduce allocation in `force` method.

julienrf committed:
Fix compilation error after rebase

julienrf committed:
Merge pull request #436 from julienrf/views

Add MapView and SeqView

julienrf committed:
Merge pull request #425 from lrytz/stringBuilderSpecific

specific C for StringBuilder

szeiger committed:
Merge pull request #476 from scala/wip/untagged-arrayseq-factory

Use WrappedArray’s iterableFactory for ArraySeq

tim-ruhland committed:
Merge branch 'master' into issue-401

szeiger committed:
Take ArrayOps out of the collection hierarchy

All collection-related methods for arrays are available on WrappedArray
so there is no need to duplicate them in ArrayOps. This allows ArrayOps
to be implemented as a value class with no inherited methods, thus it
should never be necessary to allocate an instance of it.

Unlike in the old collection library `ArrayOps` is not specialized. The
methods that were specialized before don’t seen worth specializing. In
particular, `apply` and `update` which *should* be specialized on arrays
are not provided at all by the new `ArrayOps` because they are available
natively on `Array`. The old specialization scheme would have no benefit
but a high cost: it is incompatible with a pure value class. Almost any
ArrayOps call in the old library needs to allocate an instance, with the
only exceptions being `length`, `apply` and `update`.

New and optimized implementations of many collection methods are
provided in ArrayOps, often adapted from Slick’s `ConstArray` class.

szeiger committed:
Make copyToArray useful

- `WrappedArray` and `ImmutableArray` can implement it with
  `Array.copy`.

- In `ArrayBuilder.addAll` we do not need to take specialization of
  `ImmutableArray` into account. The old implementation used
  `Array.copy` which falls back to a slow copy when the element types
  don’t match. This is still faster than the primitive implementation in
  `super.addAll` and `Iterator.copyToArray` as the last resort at least
  should not be worse. The new `addAll` based on `copyToArray` works
  well with all collection types that implement `copyToArray`
  efficiently.

- `prependedAll` and `appendedAll` in both, `ImmutableArray` and
  `ArrayOps` also do not need to optimize for known `ImmutableArray`
  arguments anymore.

szeiger committed:
Additional methods and workarounds for Scala 2.12 and Dotty issues

szeiger committed:
Merge pull request #474 from scala/wip/fix-set-plus-variance

Make deprecated + work for generic Set and Map

szeiger committed:
Introduce IterableOnceOps as common abstraction for Iterable & Iterator

- Unlike the other Ops types, IterableOnceOps is *not* implemented by
  IterableOnce but only by Iterator and Iterable.

- Fix the hierarchy of MultiSets and SortedMultiSets: The refactoring
  caused soundness problems to show up that were not reported before.
  These types now follow the usual diamond pattern where
  mutable/immutable extends generic and sorted extends unsorted.

szeiger committed:
Avoid clearing arrays unnecessarily on copy

marcelocenerine committed:
Override HashMap.default in Seq.occCounts instead of using Map.withDefaultValue

szeiger committed:
Merge pull request #457 from ruhly/issue-401

add the bare and separator-only variants of addString to Iterator

julienrf committed:
Merge pull request #482 from marcelocenerine/occCounts

 Override HashMap.default in Seq.occCounts instead of using Map.withDefaultValue

optician committed:
Fixes #382 - several entries with the same key in mutable.HashMap (#484)

* * Add HashMap test to check https://github.com/scala/collection-strawman/issues/382
* Clean code.
* Repeat search of key after evaluating default value.

* * Fix code style.

szeiger committed:
Merge pull request #480 from scala/wip/iterableonceops

Introduce IterableOnceOps as common abstraction for Iterable & Iterator

szeiger committed:
Merge pull request #479 from scala/wip/collection-less-arrayops

Don't extend collection types in ArrayOps

NthPortal committed:
Add size methods to VectorBuilder

Fixes scala/bug#9904

isnotinvain committed:
Make MapWrapper.Entry's hashCode conform to the contract in java.util.Map.Entry's documentation

tim-ruhland committed:
Fixed mapInPlace on mutable maps and sets (#459)

* Fixed mapInPlace on mutable maps and sets

* updated to provide working mapInPlace

* implements requested change to provide working mapInPlace

* added tests and scaladoc for mapInPlace on mutable Maps and Sets

szeiger committed:
Fix some ArrayOps bugs

Discovered by the compiler test suite

julienrf committed:
Merge pull request #487 from isnotinvain/issue/10663

Make MapWrapper.Entry's hashCode conform to the contract in java.util.Map.Entry's documentation

szeiger committed:
Merge pull request #488 from scala/wip/fix-arrayops

Fix some ArrayOps bugs

szeiger committed:
Remove Traversable

It was only there to allow bootstrapping on top of 2.13.0-M3. It is no
longer necessary after https://github.com/scala/scala/pull/6364.

szeiger committed:
Improve ImmutableArray operations

Now that we have efficient Array operations without wrapping in
ArrayOps we can use them in ImmutableArray

szeiger committed:
Merge pull request #486 from NthPortal/bug#9904/PR

Add size methods to VectorBuilder

julienrf committed:
Introduce typeclasses making it easier to write generic extension methods (#478)

* Introduce HasSeqOps
* Introduce HasImmutableMapOps
* Introduce HasIterableOps

allanrenucci committed:
Update to Dotty 0.7.0-RC1

julienrf committed:
Add back prependAll to buffer. Uncomment some ArrayDeque tests.

julienrf committed:
Constructing a mutable BitSet with an empty array will return an empty instance.

Fixes scala/bug#10399

Backport of 34d2f332d48adfd77c3060f948197326528b40d9 (by @joymufeng)

julienrf committed:
Complete scaladoc of Iterator

Backport of 58be110785aaef46333d59856c5f12cb575f4159 (by @Philippus)

julienrf committed:
Override getOrElse in immutable.MapX to avoid allocations.

Backport of 41133d93f78911bc60ce5e7ce7c9a242ce1b23a8 (by @id-ilych)

julienrf committed:
Document take and drop in Iterable

Backport of e29bff586f4c819e5b1f03ed8a497ac4a245f5fa (by @markus1189)

julienrf committed:
Document thrown exception of min/may(By) on empty collections

Backport of 55541745edcbb7b1d5a9830bad880cf5f543c0fe (by @markus1189)

julienrf committed:
Add more tests for Iterator#range

Backport of 836297f305bd6901b60c7410bce1be25725b6ce3 (by Steve Robinson)

julienrf committed:
Optimize tails in LinearSeqOps

Backport of 7de41a40ecb0c5fbb8450823dc08218d93f4dc81 (by Håkon Hjelde Wold)

julienrf committed:
Improve scaladoc link refs and syntax

Backport of ea65e04beef05708f0d11377338605377c86d3b6 (by @ashawley)

julienrf committed:
Improve scaladocs on stability of sort methods

Backport of 37dd50c0f1204354ec147b445652cd725126e063 (by @DanielMoss)

julienrf committed:
Add test on laziness of Iterator#scanLeft.

Backport of 5c7b14219b746812089c39876edaba9ee226a4bb (by @som-snytt).
Note that our implementation was already lazy enough, so I didn’t copy
the implementation of the referenced commit.

julienrf committed:
Add test checking that we don’t evaluate the head of an iterable twice

Backport of 0ee887573512ecd6411ac656ca03fc43696fa710 (by @martijnhoekstra)

julienrf committed:
Merge pull request #491 from dotty-staging/dotty-0.7

Update to Dotty 0.7.0-RC1

julienrf committed:
Remove LinkedList and MutableList

julienrf committed:
Add implicit BuildFrom and Factory instances for BitSets

Fixes scala/collection-strawman#438

Also removes a unused implicit conversion from IterableFactory
to BuildFrom.

szeiger committed:
Merge pull request #492 from scala/wip/immutable-array-ops

Improve ImmutableArray operations

julienrf committed:
Merge pull request #495 from julienrf/buildfrom-bitset

Add implicit BuildFrom and Factory instances for BitSets

julienrf committed:
Override knownSize in empty collections (#496)

* Override knownSize in empty collections

Fixes scala/collection-strawman#437

* Add special treatment in indexOfSlice in case the slice is empty

julienrf committed:
Make transformation operations of WeakHashMap return WeakHashMaps

Fixes scala/collection-strawman#402

julienrf committed:
More specific return type for transformation operations of IntMap, LongMap and AnyRefMap

Fixes scala/collection-strawman#403

odd committed:
Add explicit result type to Builder.mapResult.

szeiger committed:
Add missing methods to ArrayOps

Some of them need to be there because the ones on WrappedArray have
the wrong types. Others can be easily optimized.

The type parameters on `Seq.diff` and `Seq.intersect` from the old
collections library are no longer necessary.

Fixes https://github.com/scala/collection-strawman/issues/501.

szeiger committed:
Merge pull request #502 from odd/builder-mapresult-with-result-type

Add explicit result type to Builder.mapResult.

julienrf committed:
Merge pull request #499 from julienrf/map-trans

More specific return type for transformation operations of IntMap, LongMap and AnyRefMap

julienrf committed:
Merge pull request #498 from julienrf/stronger-weakhashmap

Make transformation operations of WeakHashMap return WeakHashMaps

julienrf committed:
Merge pull request #490 from julienrf/array-dequeue-squashed

Add ArrayDequeue

julienrf committed:
Merge pull request #503 from scala/wip/more-arrayops-methods

Add missing methods to ArrayOps

julienrf committed:
Move documentation of take and drop in IterableOnceOps

julienrf committed:
Constrain type parameters of LinearSeqOps

szeiger committed:
Merge pull request #493 from julienrf/fixes-from-scala-scala

Fixes from scala/scala repository

julienrf committed:
Add unzip3 to IterableOps (#497)

* Add unzip3 to IterableOps

Fixes scala/collection-strawman#433

szeiger committed:
Fix bounds checks, etc. in ArrayDeque

This also affects subclasses like mutable.Queue.

- Remove unnecessary `from` paramter in `requireBounds`.
- Allow `copyToSlice` directly beyond the end of a target array for a
  no-op. This is required to make it work for empty arrays.
- Allow `insert` and `insertAll` directly beyond the end of the
  `ArrayDeque` for appending to the end.
- Remove explicit `className` in `ArrayDeque` to have it computed
  correctly for subclasses like `Queue`.

Fixes https://github.com/scala/collection-strawman/issues/506

julienrf committed:
Merge pull request #513 from scala/wip/fix-arraydeque-bounds

Fix bounds checks, etc. in ArrayDeque

julienrf committed:
Merge pull request #511 from scala/wip/remove-traversable

Remove Traversable

julienrf committed:
Merge pull request #512 from julienrf/fix-lazylist-tostring-squashed

Make toString in LazyList consistent with other collections.

szeiger committed:
More fixes from latest rebasing of Scala 2.13.x integration build

- Restore `IterableFactory.toBuildFrom` which was erroneously removed
  in https://github.com/scala/collection-strawman/pull/495
- Fix `ArrayOps.ReverseIterator`
- More explicit syntax in `AnyRefMap.collect` implementation to make it
  compile with and without Predef
- Add a deprecated `front` method to `mutable.Queue`

julienrf committed:
Mention in the README that new PRs should be addressed to the scala/scala repository

julienrf committed:
Add tests for toBuildFrom implicit conversions

szeiger committed:
More fixes from latest rebasing of Scala 2.13.x integration build

- Change Seq-valued methods in ArrayDeque and Queue to return an
  immutable.Seq (which they used to build anyway).
- Add mutable.Stack.top and undeprecate mutable.Queue.front. Keeping
  these methods around is useful to avoid relying too much on the
  internal design. In particular, the top element of a mutable.Stack
  is the last one, not the first, when viewed as a Seq.
- Make StringOps.s private. We can never accidentally end up with a
  StringOps (or ArrayOps) in the new design so there is no need to
  unwrap (using implicits which were previously defined in Predef).
- Add StringBuilder companion object.
- Add ArrayStack aliases.

julienrf committed:
Merge pull request #516 from julienrf/readme-update

Mention in the README that new PRs should be addressed to the scala/scala repo

szeiger committed:
Reverse the direction of mutable.Stack

The top element is now at index 0, like in the old Stack and ArrayStack
implementations.

szeiger committed:
Merge pull request #514 from scala/wip/compat4

More fixes from latest rebasing of Scala 2.13.x integration build

szeiger committed:
Don’t expose Set{1..4} names via className / toString

They all print as “Set” now, like in the old collection library.

szeiger committed:
Required changes to build with new collection library

- Source code changes in all modules required for publish-core

- Adapt all JUnit tests to the new library

- Use a scaladoc stub when building in publish-core so we can build a
  compiler that satisfies sbt’s compiler interface but doesn’t require
  scala-xml to bootstrap

szeiger committed:
Merge pull request #518 from scala/wip/fix-set-names

Don’t expose Set{1..4} names via className / toString

szeiger committed:
Use temporary repository for bootstrapping

szeiger committed:
Migrate collection-strawman into standard library

This commit is the result of a scripted migration from the collection-strawman
repository into the main Scala repository. The parent commit is
5b973002e3fff1e7ba65ade82617c21cb37d059f in the master branch of
https://github.com/scala/collection-strawman.git.

The merge commit performs the following changes:
- Move the main strawman sources into the scala.collection namespace under
  src/library/scala/collection. The necessary migration steps have been
  performed and the sources should be fully functional.
- Move the tests to test/collection-strawman. They still need to be integrated
  into the standard test suite in a manual step.
- Delete all other parts (benchmarks, scalafix rules, documentation,
  collections-contrib project) of collection-strawman. They will be moved to
  other repositories.

The migration was performed by the following script:

#!/bin/bash

mkdir migration-tmp
cd migration-tmp

# Clone required repositories
git clone --no-checkout https://github.com/scala/scala.git
cd scala
git remote add strawman https://github.com/scala/collection-strawman.git
git fetch strawman master
git remote add szeiger https://github.com/szeiger/scala.git
git fetch szeiger wip/integrate-new-collections
ORIGINAL_STRAWMAN_SHA=`git rev-parse refs/remotes/strawman/master`

# Perform package migration and store files to keep in ../collection and ../test
git checkout --detach refs/remotes/strawman/master
cd collections/src/main/scala/strawman/collection
cat >migrate-source.sh <<"EOF"
echo migrating "$1"
mkdir -p ../../../../../../../collection/`dirname "$1"`
sed -E \
  -e 's/strawman\.collection/scala.collection/g' \
  -e 's/package strawman/package scala/g' \
  -e 's/private\[strawman\]/private[scala]/g' \
  -e 's/\.toStrawman//g' \
  -e 's/\.toClassic//g' \
  -e 's|//@migration|@scala.annotation.migration|g' \
  <"$1" | egrep -v \
  -e '^(import scala\.[^.]*)$' \
  -e '^(import scala\.Predef\.[^.]*)$' \
  >tmpfile
if [ "$1" = "./package.scala" ]; then
  sed -e 's/implicit //g' <tmpfile >"../../../../../../../collection/$1"
else
  mv tmpfile "../../../../../../../collection/$1"
fi
EOF
chmod 755 migrate-source.sh
find . -iname "*.scala" -exec ./migrate-source.sh "{}" \;
find . -iname "*.java" -exec ./migrate-source.sh "{}" \;
rm migrate-source.sh
cd ../../../../../../
mv test ../

# Create a wip branch from 2.13.x with the merge commit
git checkout -f -b wip/integrate-collection-strawman origin/2.13.x
git merge -s ours --allow-unrelated-histories --no-commit refs/remotes/strawman/master
rm -r src/library/scala/collection
mv ../collection src/library/scala/
mv ../test test/collection-strawman
git add src test
cat >../commitmsg <<EOF
Migrate collection-strawman into standard library

This commit is the result of a scripted migration from the collection-strawman
repository into the main Scala repository. The parent commit is
$ORIGINAL_STRAWMAN_SHA in the master branch of
https://github.com/scala/collection-strawman.git.

The merge commit performs the following changes:
- Move the main strawman sources into the scala.collection namespace under
  src/library/scala/collection. The necessary migration steps have been
  performed and the sources should be fully functional.
- Move the tests to test/collection-strawman. They still need to be integrated
  into the standard test suite in a manual step.
- Delete all other parts (benchmarks, scalafix rules, documentation,
  collections-contrib project) of collection-strawman. They will be moved to
  other repositories.

The migration was performed by the following script:

EOF
cat >>../commitmsg <$0
git commit -F ../commitmsg

# Cherry-pick the commit that makes everything work
git cherry-pick refs/remotes/szeiger/wip/integrate-new-collections

lrytz committed:
xml and partest versions that compile with 2.13 collections

lrytz committed:
Remove scaladoc stub

lrytz committed:
Temporary fix for ListBuffer.update

lrytz committed:
Build a new starr first

We need 3 build stages, locker+quick is not enough, because of scaladoc
and xml. Say we build scala `vLOCKER` with `vSTARR`, then the
compilation classpath when compiling scaladoc contains `xml_vSTARR`. The
`compiler.pom` will depend on `xml_vSTARR`.

When we build quick, the `doc` command will run `scaladoc-vLOCKER` with
`xml_vSTARR` on the runtime classpath. This is broken because
`library-vLOCKER` might not provide what `xml_vSTARR` expects. In the
the case of the new collections, `scala.collection.SeqLike` is missing.

Note that this issue shows that we need to use the same version number
for quick and locker, otherwise the quick `compiler.pom` will depend on
the wrong version of xml.

lrytz committed:
Streamline the bootstrap script

Ensure that modules are built in each stage (if they need building),
including even starr. This ensures that the locker scaladoc when
building quick runs with a freshly built xml on the classpath, which
can be necessary for binary compatibility.

Run scaladoc only in the quick stage (not starr or locker), for scala
and the modules.

Don't use `clone --reference` to build starr, doesn't work with shallow
clones (on travis).

Set `STARR_VER` to the full `SCALA_VER` + a suffix, so that it gets
the same treatment in sbt with respect to cross-versioning (binary
vs full).

The quick stage either publishes to scala-integration or sonatype
(for releases).

retronym committed:
Override List.equals for performance

retronym committed:
More overrides for performance

retronym committed:
Prefer new Array for non-generic Array instantiation

retronym committed:
Avoid indirection through the coll method

lrytz committed:
Update starr to internal release with new collections

julienrf committed:
Fix t3346e

- A fallback BuildFrom instance from String to IndexedSeq was missing
- Priorities were not right between BuildFrom instances for BitSet
  and SortedSet

lrytz committed:
Fix Queue.dequeueFirst

julienrf committed:
Fix/Update enrich-gentraversable.scala.

A better way to implement such enrichments should be shown in
collections-contrib.

lrytz committed:
Specify BitSet.className

Avoids leaking `BitSet1`.

lrytz committed:
Fix slice for negative bounds

Same fix as c05b850cf5, this time for the new collections.

lrytz committed:
Fix appending non-Queue to Queue

lrytz committed:
Specify HashSet.className

lrytz committed:
Use correct seed for ImmutableArray.hashCode

This makes the hash code consistent with other collection types.

lrytz committed:
Remove t4332.scala

This test no longer applies to the design of the new collections.
It checked that in old collections, all operations are overridden
in View subclasses to ensure lazyness. We get this by design now.

lrytz committed:
GroupedIterator extends Iterator[immutable.Seq] (not just collection.Seq)

lrytz committed:
Tighten signature of ImmutableArray.unsafeWrapArray

Otherwise, type inference will compute `ImmutableArray[Nothing]` in
cases where there's no expected type. Before this change:

    scala> collection.immutable.ImmutableArray.unsafeWrapArray(Array(1)).head
    java.lang.ClassCastException: java.lang.Integer cannot be cast to scala.runtime.Nothing$

The new signature requires additional casts at the use-site. The tricky
case is when there's an array of boxed primitives, and the user would
like to get an `ImmutableArray[Int]`. The following crashes:

    scala> val a: Array[Any] = Array(1)
    a: Array[Any] = Array(1)

    scala> collection.immutable.ImmutableArray.unsafeWrapArray(a.asInstanceOf[Array[Int]])
    java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [I

Instead, the user has to cast the resulting `ImmutableArray`:

    scala> collection.immutable.ImmutableArray.unsafeWrapArray(a).asInstanceOf[collection.immutable.ImmutableArray[Int]]
    res3: scala.collection.immutable.ImmutableArray[Int] = ImmutableArray(1)

lrytz committed:
Update IsIterableLike.arrayRepr

The previous version no longer worked because ArrayOps is no longer
part of the collections hierarchy.

lrytz committed:
Consolidate tests for invalid lubs

lrytz committed:
partest --run passes

lrytz committed:
Disable unstable part of neg/variances-refinement.scala

lrytz committed:
BuildFrom instance for Iterable

lrytz committed:
Specify ListSet.className

lrytz committed:
Bring back StringBuilder.appendAll(IterableOnce)

lrytz committed:
partest --jvm passes

lrytz committed:
Tighten signature of WrappedArray.make

szeiger committed:
Integrate JUnit tests from collection-strawman

There was a lot of duplication because they often started out as
copies of existing tests from scala/scala, adapted to the new package
structure. I have ported all new tests back into the main JUnit test
suite.

lrytz committed:
Update test/files/speclib/instrumented.jar

The file contains patched versions of ScalaRunTime and BoxesRunTime.
The jar can be re-generated by the mkinstrumented.sh, then needs to
be uploaded to bintray.com/typesafe/scala-sha-bootstrap.

Updated the patches that the script applies to use the unified format,
which provides some context, and is less likely to break in the future.

Restored the `arrayUpdateCount` counter. It was removed in a19babc952,
but the specialized tests actually use it. Note that no instrumented.jar
was ever published to bintray with the changes in a19babc952.

partest --specialized passes (also --res, --scalap, --instrumented)

lrytz committed:
Avoid infinite recursion on interactive.Global.ResponseMap

The call to `+=` in `override def addOne { super.+= }` causes an
infinite recursion (as `+=` calls `addOne`).

partest --presentation passes

lrytz committed:
'partest --srcpath scaladoc' passes

lrytz committed:
scalacheck/test passes

Released scalacheck `1.14.0-newCollections` built with Scala
`2.13.0-M4-pre-20d3c21` under the `org.scala-lang.modules` group id.
https://github.com/lrytz/scalacheck/releases/tag/v1.14.0-newCollections

Procedure:
  - `export SONATYPE_USERNAME=lamp`
  - `export SONATYPE_PASSWORD=...`
  - sbt clean test jvm/publishSigned

Adding the sonatype staging resolver for now, until the release
actually shows up on maven central.

lrytz committed:
Disable OSGi tests for now

lrytz committed:
Make Iterator.++ final (it's marked inline)

lrytz commented:
Jenkins and Travis are happy with it 🍻

lrytz committed:
Remove sonatype-staging resolver, the custom scalacheck is now on maven

lrytz committed:
Ensure stability in `Typer.packedType`

Method `packedType` does an existential over local symbols. The locals
are collected in a Set, which is converted `toList`. This is not stable,
but didn't show up in the old collections because of the Set1/Set2
special cases. In the ChampHashSet it started showing for small sets.

This makes `neg/variances-refinement.scala` stable again.

adriaanm commented on src/compiler/scala/tools/nsc/transform/Delambdafy.scala:
```diff
@@ -347,7 +347,7 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
 
     // the set of lifted lambda body methods that each method refers to
-    private val liftedMethodReferences = mutable.Map[Symbol, Set[Symbol]]().withDefault(_ => mutable.Set())
+    private val liftedMethodReferences = mutable.Map[Symbol, mutable.Set[Symbol]]()
```
No more `withDefault`?

szeiger replied:
See the changes a few lines further down:

```
-        if (currentMethod.exists) liftedMethodReferences(currentMethod) += sel.symbol
+        if (currentMethod.exists) liftedMethodReferences.getOrElseUpdate(currentMethod, mutable.Set()) += sel.symbol
```

This did the right thing only by accident with the old collections (but doesn't anymore with the new ones). The intention is apparently to update the mutable set but a default is computed from scratch on every access, it is never stored back into the map automatically, i.e. `liftedMethodReferences(currentMethod) += sel.symbol` computes an empty mutable set as the default value, adds the symbol to it, and then discards the result. It worked with the old collections because `+=` actually desugared to a `+` for the mutable set (thus computing a new mutable set instead of updating in place) and an `update_=` for the map.

We still have `withDefault` and could use it here (while fixing only the calls to update the sets and map) but not doing so makes the code clearer.

adriaanm replied:
I agree the old code was unclear (trying to be too clever). I'm curious to understand why it stopped working in the new collections. Is it because there's a `+=` method on immutable `Set` now? Or what is the precise change that broke this?

szeiger replied:
Note that it's only typed as `collection.Set`, not `mutable.Set`. We removed `collection.Set.+` to prevent users from accidentally calling it on mutable sets. Since the code looked like it should call `+=` the obvious change that I did initially was to change the type to `mutable.Set` which led to the bug because it didn't really call `+=` before.

We have since added back a deprecated version of `collection.Set.+` so that the original code would compile (and work the same way it did in 2.12).

adriaanm replied:
Cool, thanks for explaining!

adriaanm commented on src/compiler/scala/tools/nsc/transform/Delambdafy.scala:
```diff
@@ -365,7 +365,7 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
           seen += symbol
           (thisReferringMethods contains symbol) ||
-            (liftedMethodReferences(symbol) exists loop) && {
+            (liftedMethodReferences.contains(symbol) && liftedMethodReferences(symbol).exists(loop)) && {
```
Could you elaborate on this change?

szeiger replied:
We don't have an empty set as the default value anymore so we have to check first if there is a (non-empty) set.

adriaanm replied:
Is there a way to only look up `symbol` once?

julienrf replied:
Yes, we can call `exists` on `liftedMethodReferences`.

szeiger replied:
`exists` can't do a hash lookup, so it will be much slower for large sets. The new implementation performs the hash lookup twice when the symbol is found which is a bit slower than before but it doesn't have to instantiate and traverse a new `mutable.Set` when the symbol is not found which should be considerably faster. Unfortunately I don't know how common these two cases are in a typical codebase. Another option would be to keep a single empty set around as a dummy value that could be used with `getOrElse` (instead of `apply`) or use `get` and live with the boxing overhead.

adriaanm replied:
Ok, I think the new version is more clear. If there's a performance problem (seems unlikely), we can fix later.

szeiger committed:
Integrate collection strawman scalacheck tests into main test suite

Like in the case of the JUnit tests, most tests in collection-strawman
were based on existing tests in scala/scala, adapted to the new package
structure, so there are very few actual changes to merge back.

adriaanm commented:
💯 for preserving as much as possible of the test suite. I know that was a lot of (tedious) work. I'm fine with all deleted tests.

adriaanm commented:
👏 👏 👏

adriaanm committed:
Merge branch '2.13.x' into 2.13.x-new-collections

adriaanm commented:
🍰
